### PR TITLE
Added CloudProvider to simulator and fixed a few issues with CloudPro…

### PR DESCRIPTION
**Pull request by kevinylee1991**
…vider UI

### CODE REVIEW

File: src/app/core/createAddComponents.js [hunk 1]
```diff
@@ -40,7 +40,7 @@ const createAddComponents = options => {
           await initFn(this.props)
         }
         const created = await (createFn || crudActions.create)({ data, context, setContext })
-        setContext({ [dataKey]: [...existing, ...created] })
+        setContext({ [dataKey]: [...existing, created] })
```
Comment (brennancheung) [inline src/app/core/createAddComponents.js]: Wow, surprised I didn't notice this earlier.

File: src/app/core/createUpdateComponents.js [hunk 2]
```diff
@@ -40,17 +40,17 @@ const createUpdateComponents = options => {
...
         const existing = await loaderFn({ setContext, context })
         if (initFn) {
           // Sometimes a component needs more than just a single GET API call.
           // This function allows for any amount of arbitrary initialization.
           await initFn(this.props)
         }
-        const updated = await updateFn({ data, context, setContext })
-        const uid = data[uniqueIdentifier]
-        const updatedList = existing.map(x => x[uniqueIdentifier] !== uid ? x : updated)
+        const updated = await updateFn({ id, data, context, setContext })
```
Comment (brennancheung) [inline src/app/core/createUpdateComponents.js]: Do we need to extract the `id` at all?  Wondering if the `updateFn` can do this on its own with what is provided in `data`.

File: src/app/plugins/kubernetes/components/infrastructure/actions.js [hunk 3]
```diff
@@ -16,23 +16,16 @@ export const loadCloudProviders = async ({ context, setContext, reload }) => {
 }
 
 export const createCloudProvider = async ({ data, context, setContext }) => {
```
Comment (brennancheung) [inline src/app/plugins/kubernetes/components/infrastructure/actions.js]: `async`, `setContext`, and `return` are unnecessary.

`export const createCloudProvider = ({ data, context }) => context.apiClient.qbert.createCloudProvider(data)`

File: src/app/plugins/kubernetes/components/infrastructure/actions.js [hunk 4]
```diff
@@ -16,23 +16,16 @@ export const loadCloudProviders = async ({ context, setContext, reload }) => {
...
+  return context.apiClient.qbert.createCloudProvider(data)
 }
 
-export const updateCloudProvider = async ({ data, context, setContext }) => {
-  const { id } = data
-  const existing = await context.apiClient.qbert.getCloudProviders()
-  const updated = await context.apiClient.qbert.updateCloudProvider(id, data)
-  const newList = existing.map(x => x.id === id ? x : updated)
-  setContext({ routers: newList })
+export const updateCloudProvider = async ({ id, data, context, setContext }) => {
```
Comment (brennancheung) [inline src/app/plugins/kubernetes/components/infrastructure/actions.js]: Same here

File: src/server/models/qbert/CreateModel.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    },
+
+    list: (context) => {
+      return loaderFn ? loaderFn(context[dataKey]) : context[dataKey]
+    },
+
+    update: (id, data, context) => {
+      const mappedData = mappingFn ? mappingFn(data, context) : data
+      const collection = context[dataKey]
+      const index = collection.findIndex(x => x[uniqueIdentifier] === id)
```
Comment (brennancheung) [inline src/server/models/qbert/CreateModel.js]: No need to figure out `index`.  Better to use `map(x => x[uniqueIdentifier] === id ? mappedData : x)`.

File: src/app/core/createUpdateComponents.js (hunk 2 above)
Comment (kevinylee1991) [reply src/app/core/createUpdateComponents.js]: I think we can get it from both, I can change it to get it from the returned object

File: src/app/plugins/kubernetes/components/infrastructure/actions.js [hunk 6]
```diff
@@ -15,24 +15,15 @@ export const loadCloudProviders = async ({ context, setContext, reload }) => {
...
 }
 
-export const createCloudProvider = async ({ data, context, setContext }) => {
-  const created = await context.apiClient.qbert.createCloudProvider(data)
-  const existing = await context.apiClient.qbert.getCloudProviders()
-  setContext({ cloudProviders: [ ...existing, created ] })
-  return created
-}
+export const createCloudProvider = ({ data }) =>
+  context.apiClient.qbert.createCloudProvider(data)
```
Comment (brennancheung) [inline src/app/plugins/kubernetes/components/infrastructure/actions.js]: Still need `context`.  Isn't it undefined now?

**Pull request merged**